services/QuillLMS/app/controllers/snapshots_controller.rb

```diff
@@ -0,0 +1,193 @@
+    {value: "1", name: "1st"},
+    {value: "2", name: "2nd"},
+    {value: "3", name: "3rd"},
+    {value: "4", name: "4th"},
+    {value: "5", name: "5th"},
+    {value: "6", name: "6th"},
+    {value: "7", name: "7th"},
+    {value: "8", name: "8th"},
+    {value: "9", name: "9th"},
+    {value: "10", name: "10th"},
+    {value: "11", name: "11th"},
+    {value: "12", name: "12th"},
+    {value: "University", name: "University"},
+    {value: "Other", name: "Other"}
+  ]
+
+  before_action :validate_request, only: [:count]
+  before_action :authorize_request, only: [:count]
+
+  def count
+    response = Rails.cache.read(cache_key)
+
+    return render json: response if response
+
+    previous_timeframe_start, current_timeframe_start, timeframe_end = calculate_timeframes
+
+    Snapshots::CacheSnapshotCountWorker.perform_async(cache_key,
+      snapshot_params[:query],
+      current_user.id,
+      {
+        name: snapshot_params[:timeframe],
+        previous_start: previous_timeframe_start,
+        current_start: current_timeframe_start,
+        current_end: timeframe_end
+      },
+      snapshot_params[:school_ids],
+      snapshot_params[:grades])
+
+    render json: { message: 'Generating snapshot' }
+  end
+
+  def options
+    render json: {
+      timeframes: TIMEFRAMES.map { |timeframe| format_timeframe_option(timeframe) },
+      schools: Snapshots::SchoolsOptionsQuery.run(current_user.id),
+      grades: GRADE_OPTIONS
+    }
+  end
+
+  private def validate_request
+    return render json: { error: 'timeframe must be present and valid' }, status: 400 unless timeframe_param_valid?
+
+    return render json: { error: 'school_ids are required' }, status: 400 unless school_ids_param_valid?
+  end
+
+  private def timeframe_param_valid?
+    TIMEFRAMES.map { |t| t[:value] }.include?(snapshot_params[:timeframe])
+  end
+
+  private def school_ids_param_valid?
+    snapshot_params[:school_ids]&.any?
+  end
+
+  private def authorize_request
+    schools_user_admins = current_user.administered_schools.pluck(:id)
+
+    return render json: { error: 'user is not authorized for all specified schools' }, status: 403 unless snapshot_params[:school_ids]&.all? { |param_id| schools_user_admins.include?(param_id.to_i) }
+  end
+
+  private def cache_key
+    previous_timeframe_start, current_timeframe_start, timeframe_end = calculate_timeframes
+
+    Snapshots::CacheKeys.generate_key(snapshots_params[:query],
+      previous_timeframe_start,
+      current_timeframe_start,
+      timeframe_end,
+      snapshot_params.fetch(:school_ids, []),
+      snapshot_params.fetch(:grades, []))
+  end
+
+  private def format_timeframe_option(timeframe)
+    {
+      value: timeframe[:value],
+      name: timeframe[:name],
+      default: (DEFAULT_TIMEFRAME == timeframe[:value])
+    }
+  end
+
+  private def find_timeframe(timeframe_value)
+    TIMEFRAMES.find { |timeframe| timeframe[:value] == timeframe_value }
```

Comment:
This can't currently be `nil` because the `before_action` `validate_request` will return an error if a timeframe is not provided and is not in the list of valid timeframes.